### PR TITLE
Support Create Trigger with System Trigger in Oracle

### LINKED ISSUE #27119: Support parsing Oracle CREATE TRIGGER sql
# Background
Hi community,
This issue is for #26878.

ShardingSphere parser engine helps users parse a SQL to get the AST (Abstract Syntax Tree) and visit this tree to get SQLStatement (Java Object). Currently, we are planning to enhance the support for Oracle SQL parsing in ShardingSphere.

More details:
https://shardingsphere.apache.org/document/current/en/reference/sharding/parse/

# Task
This issue is to support more oracle sql parse, as follows:
```sql
CREATE TRIGGER log_errors
  AFTER SERVERERROR ON DATABASE
  BEGIN
    IF (IS_SERVERERROR (1017)) THEN
      NULL;  -- (substitute code that processes logon error)
    ELSE
      NULL;  -- (substitute code that logs error code)
    END IF;
  END;
```

```sql
CREATE TEMPORARY TABLESPACE lmtemp2 TEMPFILE '/u02/oracle/data/lmtemp201.dbf'
     SIZE 50M
     TABLESPACE GROUP group1;

ALTER TABLESPACE lmtemp TABLESPACE GROUP group2;
```

```sql
CREATE TRIGGER Contacts_BI
  BEFORE INSERT ON Contacts FOR EACH ROW
BEGIN
  :NEW.ID := Contacts_Seq.NEXTVAL;
END;
```

```sql
CREATE TRIGGER adg_logon_sync_trigger
 AFTER LOGON ON user.schema
  begin
    if (SYS_CONTEXT('USERENV', 'DATABASE_ROLE')  IN ('PHYSICAL STANDBY')) then
      execute immediate 'alter session sync with primary';
    end if;
  end;
```

```sql
CREATE TRIGGER scott.emp_permit_changes
     BEFORE
     DELETE OR INSERT OR UPDATE
     ON scott.emp
     .
     .
     .
pl/sql block 
     .
     .
     .
```

# Process
1. First confirm that this is a correct oracle sql syntax, if not please ignore;
2. Compare SQL definitions in Oficial SQL Doc and ShardingSphere SQL Doc;
3. If there is any difference in ShardingSphere SQL Doc, please correct them by referring to the Official SQL Doc;
4. Run mvn install the current_file_module;
5. Check whether there are any exceptions. If indeed, please fix them. (Especially xxxVisitor.class);
6. Add new corresponding SQL case in SQL Cases and expected parsed result in Expected Statment XML;
7. Run SQLParserParameterizedTest to make sure no exceptions.

# Relevant Skills
1. Master JAVA language
2. Have a basic understanding of Antlr `g4` file
3. Be familiar with Oracle SQLs

### PR DESCRIPTION
**Pull request by ilyasahsan123**
Ref #27119.

Changes proposed in this pull request:
  - Support Create Trigger with System Trigger

---

Before committing this PR, I'm sure that I have checked the following options:
- [x] My code follows the [code of conduct](https://shardingsphere.apache.org/community/en/involved/conduct/code/) of this project.
- [x] I have self-reviewed the commit code.
- [ ] I have (or in comment I request) added corresponding labels for the pull request.
- [x] I have passed maven check locally : `./mvnw clean install -B -T1C -Dmaven.javadoc.skip -Dmaven.jacoco.skip -e`.
- [ ] I have made corresponding changes to the documentation.
- [x] I have added corresponding unit tests for my changes.

### CODE REVIEW

**Comment by ilyasahsan123** [comment]
hi @zihaoAK47, when you get a change, could you please check this out? thanks a bunch!

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 [hunk 1]
```diff
@@ -280,9 +284,37 @@ autonomousTransPragma
...
+    : (dmlEventElement (OR dmlEventElement)* ON viewName)
+    | (nonDmlEventElement (OR nonDmlEventElement)* ON (DATABASE | (schemaName DOT_)? SCHEMA))
     ;
 
 dmlEventElement
     : (DELETE | INSERT | UPDATE) (OF LP_ columnName (COMMA_ columnName)* RP_)?
     ;
+
+nonDmlEventElement
+    : ALTER
```
Comment (zihaoAK47) [inline parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4:296]: Can you provide a grammar diagram reference here?

**Review by zihaoAK47 — CHANGES_REQUESTED** [review]

**Comment by zihaoAK47** [comment]
@ilyasahsan123 Hi, if #27119 The issue is still ongoing, please Fixes #27119 Change to Ref #27119. Thank you

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 (hunk 1 above)
Comment (ilyasahsan123) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4:296]: here is the reference:
https://docs.oracle.com/cd/B13789_01/server.101/b10759/statements_7004.htm

Screenshot 2024-01-01 at 13 12 59

Please let me know, if you have any concern.

**Comment by ilyasahsan123** [comment]
> @ilyasahsan123 Hi, if #27119 The issue is still ongoing, please Fixes #27119 Change to Ref #27119. Thank you

hi @zihaoAK47, I apologize, but I'm a bit lost. Could you please clarify what you mean?

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 (hunk 1 above)
Comment (zihaoAK47) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4:296]:
```
CREATE TRIGGER log_errors
  AFTER SERVERERROR ON DATABASE
  BEGIN
    IF (IS_SERVERERROR (1017)) THEN
      NULL;  -- (substitute code that processes logon error)
    ELSE
      NULL;  -- (substitute code that logs error code)
    END IF;
  END;
```

Thank you for providing the document reference. It seems that some parts of SQL should be handled by the `system_trigger` syntax parse.

Ref: https://docs.oracle.com/en/database/oracle/oracle-database/19/lnpls/CREATE-TRIGGER-statement.html#GUID-AF9E33F1-64D1-4382-A6A4-EC33C36F237B__BABHEFGE

**Comment by zihaoAK47** [comment]
> > @ilyasahsan123 Hi, if #27119 The issue is still ongoing, please Fixes #27119 Change to Ref #27119. Thank you
> 
> hi @zihaoAK47, I apologize, but I'm a bit lost. Could you please clarify what you mean?

Fixes will close Issues. If there are still tasks to be processed inside, we need to use Ref to reference this issue

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 (hunk 1 above)
Comment (ilyasahsan123) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4:309]: Thanks for your suggestion, @zihaoAK47, i have updated the changes, could you please take a look?

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 [hunk 2]
```diff
@@ -279,10 +283,51 @@ autonomousTransPragma
...
+simpleDmlTrigger
+    : (BEFORE | AFTER) dmlEventClause
+    ;
+
+dmlEventClause
+    : dmlEventTrigger (OR dmlEventTrigger)* ON viewName
+    ;
+
+dmlEventTrigger
+    : DELETE
```
Comment (zihaoAK47) [inline parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4:299]: Please keep these branches in one line.
Refer code conduct - https://shardingsphere.apache.org/community/en/involved/conduct/code/.
> If a rule’s branch is over than 5, every branch take a new line.

**Review by zihaoAK47 — CHANGES_REQUESTED** [review]

File: parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4 (hunk 2 above)
Comment (ilyasahsan123) [reply parser/sql/dialect/oracle/src/main/antlr4/imports/oracle/PLSQL.g4]: Done. Thanks

**Review by zihaoAK47 — APPROVED** [review]
Thank you, merged

**Pull request merged**